rllib/core/optim/tests/test_rl_optimizer.py

```diff
@@ -0,0 +1,378 @@
+        """
+        return gradients_dict
+
+    def compile_results(
+        self,
+        batch: NestedDict,
+        fwd_out: Mapping[str, Any],
+        postprocessed_loss: Mapping[str, Any],
+        post_processed_gradients: Mapping[str, Any],
+    ) -> Mapping[str, Any]:
+        """Compile results from the update.
+
+        Args:
+            batch: The batch that was used for the update.
+            fwd_out: The output of the forward train pass.
+            postprocessed_loss: The loss after postprocessing.
+            post_processed_gradients: The gradients after postprocessing.
+
+        Returns:
+            A dictionary of results.
+        """
+
+        loss_numpy = convert_to_numpy(postprocessed_loss)
+        results = {
+            "avg_reward": batch["rewards"].mean(),
+        }
+        results.update(loss_numpy)
+        results["module_norm"] = model_norm(self._module)
+        return results
+
+    def update(self, batch: NestedDict) -> Mapping[str, Any]:
+        """Perform an update on this Trainer.
+
+        Args:
+            batch: A batch of data.
+
+        Returns:
+            A dictionary of results.
+        """
+        torch_batch = convert_to_torch_tensor(batch)
+        fwd_out = self._module.forward_train(torch_batch)
+        loss = self._rl_optimizer.compute_loss(torch_batch, fwd_out)
+        gradients = self.compute_gradients(loss)
+        post_processed_gradients = self.on_after_compute_gradients(gradients)
+        self.apply_gradients(post_processed_gradients)
+        results = self.compile_results(batch, fwd_out, loss, post_processed_gradients)
+        return results
+
+    def compute_gradients(
+        self, optimization_vars: Mapping[str, Any]
+    ) -> Mapping[str, Any]:
+        """Perform an update on self._module
+
+            For example compute and apply gradients to self._module if
+                necessary.
+
+        Args:
+            optimization_vars: named variables used for optimizing self._module
+
+        Returns:
+            A dictionary of extra information and statistics.
+        """
+        # for torch
+        optimization_vars["total_loss"].backward()
+        grads = {n: p.grad for n, p in self._module.named_parameters()}
+        return grads
+
+    def apply_gradients(self, gradients: Mapping[str, Any]) -> None:
+        """Perform an update on self._module"""
+        for optimizer in self._rl_optimizer.get_optimizers().values():
+            optimizer.step()
+            optimizer.zero_grad()
+
+    def set_state(self, state: Mapping[str, Any]) -> None:
+        """Set the state of the trainer."""
+        self._module.set_state(state.get("module_state", {}))
+
+    def get_state(self) -> Mapping[str, Any]:
+        """Get the state of the trainer."""
+        return {
+            "module_state": self._module.get_state(),
+            "optimizer_state": self._rl_optimizer.get_state(),
+        }
+
+
+def test_rl_optimizer_example_torch():
```

Comment:
Can we stick to the unittest.TestCase pattern in all our test cases?


```
class TestRLOptimizer(unittest.TestCase):
    def test_blabla(self):
        ...
    def test_blabla2(self):
        ...
```

This way, you can run both unittest and pytest on these scripts.

Reply:
+1. Also For the places that you want to use check to make sure two things are not equal use
```
check(x, y, false=True)
```

Reply:
"we can but just fyi, both unitest and pytest will pick up any functions that begin with the word "test" and run them as tests.
@kouroshHakha @sven1977